### ISSUE TITLE
Implement p2p related tools

### ISSUE THREAD

**Issue by int88** [paradigmxyz/reth#7411]
### Describe the feature

There are some p2p related tools geth, like:

https://geth.ethereum.org/docs/tools/devp2p

https://github.com/ethereum/go-ethereum/tree/master/p2p/simulations

If it's meaningful to implement related tools in reth, I'd like to do it!

@mattsse @emhane WDTY?

**Comment by mattsse** [comment]
sgtm!

if you want, you can add these as standalone crates first

**Comment by int88** [comment]
devp2p command TODO list:

- enrump: Pretty-prints node records
  - [ ] enrdump
- key: Operations on node keys
  - [ ] generate: Generate node key files
  - [ ] to-id: Create a node ID from a node key file
  - [ ] to-encode: Creates an encode URL from a node key file
  - [ ] to-enr: Creates an ENR from a node key file
- discv4: Node Discovery v4 tools
  - [ ] ping: Sends ping to a node
  - [ ] requestenr: Requests a node record using EIP-868 enrRequest
  - [ ] resolve: Finds a node in the DHT
  - [ ] resolve-json: Re-resolves node in a nodes.json file
  - [ ] crawl: Updates a nodes.json file with random nodes found in the DHT
  - [ ] test: Runs tests against a node
- discv5: Node Discovery v5 tools
  - [ ] ping: Sends ping to a node
  - [ ] resolve: Finds a node in the DHT
  - [ ] crawl: Updates a nodes.json file with random nodes found in the DHT
  - [ ] test: Runs test against a node
  - [ ] listen: Runs a node
- dns: DNS Discovery Commands
  - [ ] sync: Download a DNS discovery tree
  - [ ] sign: Sign a DNS discovery tree
  - [ ] to-txt: Creates a DNS TXT records for a discovery tree
  - [ ] to-cloudflare: Deploy DNS TXT records to CloudFlare
  - [ ] to-route53: Deploy DNS TXT records to Amazon Route53
  - [ ] nuke-route53: Deletes DNS TXT records of a subdomain on Amazon Route53
- nodeset: Node set tools
  - [ ] info: Show statistics about a node set
  - [ ] filter: Filters a node set
- rlpx: RLPx Commands
  - [ ] ping: ping <node>
  - [ ] eth-test: Runs eth protocol tests against a node
  - [ ] snap-test: Runs snap protocol tests against a node

**Comment by emhane** [comment]
for sure! really nice idea

**Comment by emhane** [comment]
I'd tackle each  todo as a separate new issue, adding to task list here. that way it's easier for us to review, more overseeable to give you specific hints for impl in that issue, and less merge conflicts with main for you since each topic branch would be short-lived.

a hint for discv5 already, is to use this method
https://github.com/paradigmxyz/reth/blob/ef3f67743d517b2b15692949efae4671168bccbd/crates/net/discv5/src/lib.rs#L381-L387
and then read the sigp `Discv5` public api thoroughly, should be possible to impl all commands with those existing methods
https://github.com/sigp/discv5/blob/master/src/discv5.rs

**Comment by Rjected** [comment]
> make sure to try this tool out first https://github.com/AgeManning/discv5-cli, no need to impl same thing again in reth imo

agree, as well as enr-cli: https://github.com/agemanning/enr-cli

`cargo install` makes using these really easy

**Comment by emhane** [comment]
> agree, as well as enr-cli: https://github.com/agemanning/enr-cli
> 
true this tool exists too

I updated your list @int88 based on review https://github.com/paradigmxyz/reth/pull/8403#pullrequestreview-2102458098 and https://github.com/paradigmxyz/reth/pull/8403#pullrequestreview-2102435753

**Comment by int88** [comment]
Thanks for all of your kind reviews @mattsse @Rjected @emhane .

So I would implement the commands filtered by @emhane .

One last concern, should these commands should be integrated into `reth p2p` or as a separate binary?

IMO, I agree with @Rjected , integrated into `reth p2p` maybe better :)

**Comment by emhane** [comment]
> One last concern, should these commands should be integrated into `reth p2p` or as a separate binary?
> 
yes, see the text I added to the list in bold @int88. check out how it is done for the `db` subcommand `clear` with subsubcommands `mdbx` and `static-file`

https://github.com/paradigmxyz/reth/blob/e0c22ace34e520b6276af2f54acb634363b7979f/bin/reth/src/commands/db/clear.rs#L13-L16

so for example, `p2p` would be the command, `rlpx` is a subcommand (like `clear`), and `ping` is a sub subcommand (like `mdbx`)

**Comment by emhane** [comment]
@int88 focus on rlpx

ok so, going through this in more detail, regarding the remaining rlpx commands:

- the ping command would be nice to have! which would somehow access `P2PStream::send_ping` via a new p2p sub command for rlpx https://github.com/paradigmxyz/reth/blob/e97e379e43b69e15d8dcea14d60f3d791be72328/crates/net/eth-wire/src/p2pstream.rs#L304-L307
- eth tests are run in hive using rpc, right @Rjected ?
- we don't have support for live snap sync yet.

in general, it would be nice to have a complete rlpx command, to manually send all the `EthMessages`.

**Comment by jenpaff** [comment]
not on our roadmap to re-implement devp2p
